src/Common/src/TypeSystem/Common/MethodDesc.cs

```diff
@@ -104,6 +104,34 @@ public bool Equals(MethodSignature otherSignature)
 
             return true;
         }
+
+        /// <summary>
+        /// Adapt current signature to the instantation contexts.
+        /// </summary>
+        /// <param name="typeInstantiation">Context used for adapting the signature if it involves generic type parameter.</param>
+        /// <param name="methodInstantiation">Context used for adapting the signature if it involves generic method type parameter.</param>
+        /// <returns></returns>
+        public MethodSignature InstantiateSignature(Instantiation typeInstantiation, Instantiation methodInstantiation)
+        {
+            // Instantiate parameters first. If instantiation has no effects, avoid the allocation of a new array.
+            TypeDesc[] parameters = null;
+            for (int i = 0; i < _parameters.Length; i++)
+            {
+                TypeDesc oldType = _parameters[i];
+                TypeDesc newType = oldType.InstantiateSignature(typeInstantiation, methodInstantiation);
```

Comment:
Could you please make sure that this works for cases like `class MyType<T> { void MyMethod(Dictionary<T,T> x); }`? I do not think it will handle these cases well.